Template/Component renderers for column headers and footers

gilberto-torrezan opened the pull request:
This PR also refactors the `addColumn` API to not support a String parameter for the header anymore.

ComponentRenderer specific logic was moved from Grid to make it easier to be reused in the future, and to reduce the complexity of the Grid itself.

Introduces a new super class for the Columns: AbstractColumn, with common logic for rendering templates.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/flow/2933)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/flow?pullRequest=2933) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/flow?pullRequest=2933) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/flow?pullRequest=2933) it.</sub>

ahie commented:
Reviewed 14 of 22 files at r1, 1 of 2 files at r3.
Review status: 14 of 23 files reviewed at latest revision, 15 unresolved discussions.

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/ComboBoxView.java, line 136 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-KyzoWebstXk5wxp2GwI:-KyzoWebstXk5wxp2GwJ:bn538e0) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/ComboBoxView.java#L136)):*
> comboBox.setValue(listOfSongs.get(1));

Unrelated to the the PR?

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java, line 681 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-KyzqJWHPvKrHXLJ2PwX:-KyzqJWHPvKrHXLJ2PwY:b-32t0qs) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java#L681)):*
> Label label = new Label("Name");
>                     label.getStyle().set("color", "green");
>                     label.setTitle("Name");
>                     return label;

Similar code is repeated, could refactor.

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java, line 720 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kyzpyu42-8oh3VPRxEC:-Kyzpyu42-8oh3VPRxED:brmcyi) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java#L720)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>
>         ColumnGroup<Person> addressColumnGroup = grid
>                 .mergeColumns(streetColumn, postalCodeColumn)
>                 .setHeader(new ComponentRenderer<>(() -> {
>                     Label label = new Label("Address Information");
>                     label.setTitle("Address Information");
>                     return label;
>                 }));
>         grid.mergeColumns(informationColumnGroup, addressColumnGroup);
</details>

These don't really feel relevant to the demo so I would leave them out.

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java, line 723 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kyzpn3Ax2_VUv7xSoVU:-Kyzpn3Ax2_VUv7xSoVV:bmbw3d1) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java#L723)):*
> Checkbox toggleSortable

Should these also be part of the example code?

---

*[flow-components-parent/flow-components-test/src/test/java/com/vaadin/flow/tests/components/grid/GridSortingTest.java, line 228 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kyzs2vLfGBwp02CPiKQ:-Kyzs2vMX0JlZVZ3KNPc:boxenk3) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/flow-components-test/src/test/java/com/vaadin/flow/tests/components/grid/GridSortingTest.java#L228)):*
> .setHeader("");

Not really needed.

---

*[flow-components-parent/flow-components-test/src/test/java/com/vaadin/flow/tests/components/grid/SortOrderBuildersTest.java, line 37 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-KyzrubNNOD9cCtolcs5:-KyzrubNNOD9cCtolcs6:boxenk3) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/flow-components-test/src/test/java/com/vaadin/flow/tests/components/grid/SortOrderBuildersTest.java#L37)):*
> .setHeader("");

Not really needed.

---

*[flow-components-parent/flow-components/src/main/java/com/vaadin/ui/grid/ColumnGroup.java, line 34 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kz-2vt91p_mReSugcP8:-Kz-2vt91p_mReSugcP9:b-qdgzo7) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/flow-components/src/main/java/com/vaadin/ui/grid/ColumnGroup.java#L34)):*
> <T>

Is the type parameter needed? Constructor takes a `Grid<?>`.

---

*[flow-documentation/flow-components/tutorial-flow-grid.asciidoc, line 246 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kz-3Q30ubVNADKp6qij:-Kz-3Q30ubVNADKp6qik:boryxgs) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-documentation/flow-components/tutorial-flow-grid.asciidoc#L246)):*
> ```AsciiDoc
> 
> [[components.grid.columns.header]]
> === Column Header and Footer
> ```

Could have a component example as well. Anyway, I also have a version of this tutorial :).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2933)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 14 of 23 files reviewed at latest revision, 14 unresolved discussions.

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/ComboBoxView.java, line 136 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-KyzoWebstXk5wxp2GwI:-Kz-7zYpeo0XEqTITT0e:b299zyz) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/ComboBoxView.java#L136)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary><blockquote>

> comboBox.setValue(listOfSongs.get(1));

Unrelated to the the PR?
</blockquote></details>

Totally unrelated. Reverted.

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java, line 681 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-KyzqJWHPvKrHXLJ2PwX:-Kz-88zsO7kXICVA9S_0:bct2360) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java#L681)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary><blockquote>

> Label label = new Label("Name");
>                     label.getStyle().set("color", "green");
>                     label.setTitle("Name");
>                     return label;

Similar code is repeated, could refactor.
</blockquote></details>

The intention here was to provide the same UI using the declarative (template) form and the component form. I can refactor for make an external method, but then we start having problems on how to show that code in the demo.

But changing the color of the label was the first thing I could imagine I could do. Do you have a better idea on what to showcase here? Ideas are appreciated :-)

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java, line 720 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kyzpyu42-8oh3VPRxEC:-Kz-8o6yjBOHSdN0AY-L:b-6r82d7) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java#L720)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary><blockquote>

>         ColumnGroup<Person> addressColumnGroup = grid
>                 .mergeColumns(streetColumn, postalCodeColumn)
>                 .setHeader(new ComponentRenderer<>(() -> {
>                     Label label = new Label("Address Information");
>                     label.setTitle("Address Information");
>                     return label;
>                 }));
>         grid.mergeColumns(informationColumnGroup, addressColumnGroup);

These don't really feel relevant to the demo so I would leave them out.
</blockquote></details>

Same as above - I tried to replicate the same UI using templates and components.

This code here was to show that you can set a component renderer for the group as well. But then, better ideas are appreciated =)

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java, line 723 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kyzpn3Ax2_VUv7xSoVU:-Kz-8lY2fDNH7W_ycqcH:bss4a2b) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java#L723)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary><blockquote>

> Checkbox toggleSortable

Should these also be part of the example code?
</blockquote></details>

Removed.

---

*[flow-components-parent/flow-components-test/src/test/java/com/vaadin/flow/tests/components/grid/GridSortingTest.java, line 228 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kyzs2vLfGBwp02CPiKQ:-Kz-BuBEIpQmxOrorR-R:bss4a2b) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/flow-components-test/src/test/java/com/vaadin/flow/tests/components/grid/GridSortingTest.java#L228)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary><blockquote>

> .setHeader("");

Not really needed.
</blockquote></details>

Removed.

---

*[flow-components-parent/flow-components-test/src/test/java/com/vaadin/flow/tests/components/grid/SortOrderBuildersTest.java, line 37 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-KyzrubNNOD9cCtolcs5:-Kz-BvhMopHhLB1Cm8tW:bss4a2b) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/flow-components-test/src/test/java/com/vaadin/flow/tests/components/grid/SortOrderBuildersTest.java#L37)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary><blockquote>

> .setHeader("");

Not really needed.
</blockquote></details>

Removed.

---

*[flow-components-parent/flow-components/src/main/java/com/vaadin/ui/grid/ColumnGroup.java, line 34 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kz-2vt91p_mReSugcP8:-Kz-CeCzrhG9ORNEfOrD:b5m9gns) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/flow-components/src/main/java/com/vaadin/ui/grid/ColumnGroup.java#L34)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary><blockquote>

> <T>

Is the type parameter needed? Constructor takes a `Grid<?>`.
</blockquote></details>

You're right, removed (that cause changes in the demos and tests).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2933)*
<!-- Sent from Reviewable.io -->

ahie commented:
Reviewed 3 of 22 files at r1, 1 of 3 files at r4, 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks broke.

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java, line 681 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-KyzqJWHPvKrHXLJ2PwX:-Kz2hIpTd5dOp3O2ODdR:b34ybse) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java#L681)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

The intention here was to provide the same UI using the declarative (template) form and the component form. I can refactor for make an external method, but then we start having problems on how to show that code in the demo.

But changing the color of the label was the first thing I could imagine I could do. Do you have a better idea on what to showcase here? Ideas are appreciated :-)
</blockquote></details>

True, the issue is how to display all of the code neatly if this were refactored. Not blocking.

---

*[flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java, line 720 at r2](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kyzpyu42-8oh3VPRxEC:-Kz2gs1gXzu80Vq8fXrS:b2lr4f3) ([raw file](https://github.com/vaadin/flow/blob/42e4bfbc833ec121f540f594d178ff007babc510/flow-components-parent/demo-flow-components/src/main/java/com/vaadin/flow/demo/views/GridView.java#L720)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

Same as above - I tried to replicate the same UI using templates and components.

This code here was to show that you can set a component renderer for the group as well. But then, better ideas are appreciated =)
</blockquote></details>

Just feels a bit overkill to have two levels of grouping. Might distract from the main feature that is being demoed.

---

*[flow-components-parent/flow-components/src/main/java/com/vaadin/ui/grid/Grid.java, line 1464 at r5](https://reviewable.io:443/reviews/vaadin/flow/2933#-Kz2fRqaXqhqsKth4rma:-Kz2fRqbJgqPXg8ia56i:b-c4ick1) ([raw file](https://github.com/vaadin/flow/blob/ab6cc932e2192f6079fc098612d8c5e52b0ed1a8/flow-components-parent/flow-components/src/main/java/com/vaadin/ui/grid/Grid.java#L1464)):*
> ```Java
>     }
> 
>     GridDataGenerator<T> getDataGenerator() {
> ```

Could document this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/2933)*
<!-- Sent from Reviewable.io -->
